File: backend/src/services/metricasService.js
// src/services/metricasService.js
// Métricas de NEGÓCIO — as que dizem se a plataforma está a funcionar.
//
// O Sentry já cobre o que se parte. Não cobre o que apenas deixa de acontecer,
// e é aí que uma plataforma B2B morre em silêncio: as cotações continuam a
// entrar e param de virar ordens; os pagamentos continuam a ser confirmados,
// mas cada vez mais tarde. Nada disso lança uma exceção. Não há erro nenhum
// para o Sentry apanhar — há um número que baixa durante semanas.
//
// TRÊS NÚMEROS E NÃO TRINTA. Um painel com trinta indicadores não se olha. Cada
// um destes responde a uma pergunta que alguém faria em voz alta:
//   1. Está a passar dinheiro por aqui?          → volume transacionado
//   2. O marketplace está a fechar negócio?      → cotação que vira ordem
//   3. O "pagamento garantido" está a cumprir?   → tempo até confirmação
//
// O terceiro é o mais importante e o menos óbvio. É a promessa do produto, é o
// que a conciliação automática existe para melhorar, e é o único que diz se
// valeu a pena — sem ele, KX-05 é uma funcionalidade sem forma de saber se
// resolveu alguma coisa.

const prisma = require('../config/database');

const PAGAS = ['PAGA', 'EM_EXECUCAO', 'ENTREGUE', 'RECEBIDA_CONFORME', 'RECEBIDA_COM_DIVERGENCIA', 'CONCLUIDA'];

function janela(dias = 30, agora = new Date()) {
  const de = new Date(agora);
  de.setDate(de.getDate() - dias);
  return { de, ate: agora };
}

/** Mediana e não média: um único negócio enorme desloca a média e não desloca a mediana. */
function mediana(valores) {
  if (!valores.length) return null;
  const ord = [...valores].sort((a, b) => a - b);
  const meio = Math.floor(ord.length / 2);
  return ord.length % 2 ? ord[meio] : (ord[meio - 1] + ord[meio]) / 2;
}

async function volumeTransacionado({ de, ate }) {
  const agg = await prisma.purchaseOrder.aggregate({
    where: { createdAt: { gte: de, lte: ate }, status: { in: PAGAS } },
    _sum: { totalAmount: true },
    _count: { _all: true },
  });
  return {
    ordens: agg._count._all,
    total: Number(agg._sum.totalAmount || 0),
    // O ticket médio explica o total: 10 ordens de 1M não é o mesmo negócio que
    // 1000 de 10 mil, e o total sozinho não distingue os dois.
    ticketMedio: agg._count._all ? Number(agg._sum.totalAmount || 0) / agg._count._all : 0,
  };
}

/**
 * Quantos pedidos de cotação acabam em ordem.
 *
 * Conta-se pelas cotações CRIADAS na janela, e não pelas ordens: uma cotação
 * pedida hoje pode virar ordem daqui a três semanas, e dividir ordens desta
 * semana por cotações desta semana produz um número que sobe e desce sozinho
 * sem nada ter mudado.
 */
async function conversaoDeCotacoes({ de, ate }) {
  const [pedidas, respondidas, fechadas] = await Promise.all([
    prisma.quoteRequest.count({ where: { createdAt: { gte: de, lte: ate } } }),
    prisma.quoteRequest.count({ where: { createdAt: { gte: de, lte: ate }, status: { in: ['RESPONDIDA', 'FECHADA'] } } }),
    prisma.quoteRequest.count({ where: { createdAt: { gte: de, lte: ate }, status: 'FECHADA' } }),
  ]);

  const pct = (n) => (pedidas ? Math.round((n / pedidas) * 1000) / 10 : null);
  return {
    pedidas,
    respondidas,
    fechadas,
    // Duas taxas e não uma: se as cotações não fecham, é preciso saber se é
    // porque os fornecedores não respondem ou porque respondem e não convence.
    taxaDeResposta: pct(respondidas),
    taxaDeFecho: pct(fechadas),
  };
}

/**
 * Quanto tempo entre a fatura ser emitida e o pagamento entrar.
 *
 * É a métrica que mede a promessa do produto. Separada POR CANAL de propósito:
 * é assim que se vê se a conciliação automática está a valer alguma coisa — se
 * a mediana da referência bancária não for muito menor do que a da
 * transferência manual, KX-05 não resolveu o que dizia resolver.
 */
async function tempoAteConfirmacao({ de, ate }) {
  const pagamentos = await prisma.payment.findMany({
    where: { processedAt: { gte: de, lte: ate }, status: 'PROCESSADO' },
    select: { processedAt: true, canal: true, invoice: { select: { issuedAt: true } } },
    take: 20000,
  });

  const horasPorCanal = new Map();
  for (const p of pagamentos) {
    if (!p.invoice?.issuedAt) continue;
    const horas = (new Date(p.processedAt) - new Date(p.invoice.issuedAt)) / 3600000;
    if (horas < 0) continue; // dados incoerentes não entram na conta
    if (!horasPorCanal.has(p.canal)) horasPorCanal.set(p.canal, []);
    horasPorCanal.get(p.canal).push(horas);
  }

  const todas = [...horasPorCanal.values()].flat();
  const arredondar = (h) => (h == null ? null : Math.round(h * 10) / 10);

  return {
    pagamentos: todas.length,
    medianaHoras: arredondar(mediana(todas)),
    porCanal: Object.fromEntries(
      [...horasPorCanal.entries()].map(([canal, hs]) => [canal, {
        pagamentos: hs.length,
        medianaHoras: arredondar(mediana(hs)),
      }]),
    ),
  };
}

/**
 * O que a conciliação NÃO conseguiu resolver sozinha.
 *
 * Se este número só cresce, o formato da descrição do banco mudou e ninguém
 * deu por isso — e o trabalho manual voltou todo, devagar, sem aviso.
 */
async function conciliacaoAutomatica({ de, ate }) {
  const [conciliadas, porResolver] = await Promise.all([
    prisma.linhaExtrato.count({ where: { importadaEm: { gte: de, lte: ate }, estado: 'CONCILIADA' } }),
    prisma.linhaExtrato.count({
      where: { importadaEm: { gte: de, lte: ate }, estado: { in: ['SEM_CORRESPONDENCIA', 'DIVERGENTE'] } },
    }),
  ]);
  const total = conciliadas + porResolver;
  return {
    linhas: total,
    conciliadas,
    porResolver,
    taxaAutomatica: total ? Math.round((conciliadas / total) * 1000) / 10 : null,
  };
}

async function resumo({ dias = 30 } = {}) {
  // Calculado UMA vez e reutilizado. Estava a ser validado para a janela e
  // repetido em bruto no período relatado, por isso um `dias=-5` media 30 dias
  // e dizia -5. Uma métrica que declara um período diferente daquele que mediu
  // é pior do que não existir: quem a lê compara-a com outra coisa.
  const pedido = Math.floor(Number(dias));
  const diasEfetivos = Number.isFinite(pedido) && pedido > 0 ? pedido : 30;
  const periodo = janela(diasEfetivos);
  const [volume, cotacoes, tempo, conciliacao] = await Promise.all([
    volumeTransacionado(periodo),
    conversaoDeCotacoes(periodo),
    tempoAteConfirmacao(periodo),
    conciliacaoAutomatica(periodo),
  ]);

  return {
    periodo: { de: periodo.de.toISOString(), ate: periodo.ate.toISOString(), dias: diasEfetivos },
    volume,
    cotacoes,
    tempoAteConfirmacao: tempo,
    conciliacao,
  };
}

module.exports = { resumo, volumeTransacionado, conversaoDeCotacoes, tempoAteConfirmacao, conciliacaoAutomatica, mediana };

File: backend-java/src/main/java/ao/kixima/faturacao/MetricasService.java
package ao.kixima.faturacao;

import ao.kixima.conciliacao.ConciliacaoService;
import ao.kixima.conciliacao.LinhaExtratoRepository;
import ao.kixima.payment.Payment;
import ao.kixima.payment.PaymentRepository;
import ao.kixima.payment.PaymentStatus;
import ao.kixima.po.PoStatus;
import ao.kixima.po.PurchaseOrderRepository;
import ao.kixima.quote.QuoteRequestRepository;
import ao.kixima.quote.QuoteStatus;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Espelha backend/src/services/metricasService.js — as três perguntas da
 * KIXIMA num período: quanto dinheiro passou, quantas cotações fecharam e
 * quanto tempo o dinheiro demora a ser confirmado (mediana, por canal), mais
 * a taxa de conciliação automática. Sem dados, as taxas ficam nulas em vez de
 * darem 0% — "0% de conversão" seria uma divisão por zero disfarçada.
 */
@Service
public class MetricasService {

    static final List<PoStatus> PAGAS = List.of(PoStatus.PAGA, PoStatus.EM_EXECUCAO, PoStatus.ENTREGUE,
            PoStatus.RECEBIDA_CONFORME, PoStatus.RECEBIDA_COM_DIVERGENCIA, PoStatus.CONCLUIDA);

    public record Janela(Instant de, Instant ate) {
    }

    private final PurchaseOrderRepository purchaseOrderRepository;
    private final QuoteRequestRepository quoteRequestRepository;
    private final PaymentRepository paymentRepository;
    private final LinhaExtratoRepository linhaExtratoRepository;

    public MetricasService(PurchaseOrderRepository purchaseOrderRepository, QuoteRequestRepository quoteRequestRepository,
                           PaymentRepository paymentRepository, LinhaExtratoRepository linhaExtratoRepository) {
        this.purchaseOrderRepository = purchaseOrderRepository;
        this.quoteRequestRepository = quoteRequestRepository;
        this.paymentRepository = paymentRepository;
        this.linhaExtratoRepository = linhaExtratoRepository;
    }

    public static Janela janela(int dias, Instant agora) {
        return new Janela(agora.minus(dias, ChronoUnit.DAYS), agora);
    }

    public static Double mediana(List<Double> valores) {
        if (valores == null || valores.isEmpty()) return null;
        List<Double> ord = new ArrayList<>(valores);
        ord.sort(Double::compare);
        int meio = ord.size() / 2;
        return ord.size() % 2 == 1 ? ord.get(meio) : (ord.get(meio - 1) + ord.get(meio)) / 2;
    }

    private static Double arredondar(Double h) {
        return h == null ? null : Math.round(h * 10) / 10.0;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> volumeTransacionado(Janela j) {
        List<Object[]> agg = purchaseOrderRepository.agregadoVolume(j.de(), j.ate(), PAGAS);
        long ordens = agg.isEmpty() ? 0 : ((Number) agg.get(0)[0]).longValue();
        BigDecimal soma = agg.isEmpty() || agg.get(0)[1] == null ? BigDecimal.ZERO : (BigDecimal) agg.get(0)[1];
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("ordens", ordens);
        m.put("total", soma.doubleValue());
        m.put("ticketMedio", ordens > 0 ? soma.doubleValue() / ordens : 0d);
        return m;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> conversaoDeCotacoes(Janela j) {
        long pedidas = quoteRequestRepository.countByCreatedAtBetween(j.de(), j.ate());
        long respondidas = quoteRequestRepository.countByCreatedAtBetweenAndStatusIn(j.de(), j.ate(), List.of(QuoteStatus.RESPONDIDA, QuoteStatus.FECHADA));
        long fechadas = quoteRequestRepository.countByCreatedAtBetweenAndStatusIn(j.de(), j.ate(), List.of(QuoteStatus.FECHADA));
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("pedidas", pedidas);
        m.put("respondidas", respondidas);
        m.put("fechadas", fechadas);
        m.put("taxaDeResposta", pedidas == 0 ? null : Math.round(respondidas * 1000.0 / pedidas) / 10.0);
        m.put("taxaDeFecho", pedidas == 0 ? null : Math.round(fechadas * 1000.0 / pedidas) / 10.0);
        return m;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> tempoAteConfirmacao(Janela j) {
        List<Payment> pagamentos = paymentRepository.findProcessadosNoPeriodo(j.de(), j.ate(), PaymentStatus.PROCESSADO, PageRequest.of(0, 20000));
        Map<String, List<Double>> horasPorCanal = new LinkedHashMap<>();
        for (Payment p : pagamentos) {
            if (p.getInvoice() == null || p.getInvoice().getIssuedAt() == null || p.getProcessedAt() == null) continue;
            double horas = (p.getProcessedAt().toEpochMilli() - p.getInvoice().getIssuedAt().toEpochMilli()) / 3600000.0;
            if (horas < 0) continue; // dados incoerentes não entram na conta
            String canal = p.getCanal() == null ? "null" : p.getCanal().name();
            horasPorCanal.computeIfAbsent(canal, k -> new ArrayList<>()).add(horas);
        }
        List<Double> todas = new ArrayList<>();
        horasPorCanal.values().forEach(todas::addAll);
        Map<String, Object> porCanal = new LinkedHashMap<>();
        for (Map.Entry<String, List<Double>> e : horasPorCanal.entrySet()) {
            Map<String, Object> c = new LinkedHashMap<>();
            c.put("pagamentos", e.getValue().size());
            c.put("medianaHoras", arredondar(mediana(e.getValue())));
            porCanal.put(e.getKey(), c);
        }
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("pagamentos", todas.size());
        m.put("medianaHoras", arredondar(mediana(todas)));
        m.put("porCanal", porCanal);
        return m;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> conciliacaoAutomatica(Janela j) {
        long conciliadas = linhaExtratoRepository.countByImportadaEmBetweenAndEstado(j.de(), j.ate(), ConciliacaoService.CONCILIADA);
        long porResolver = linhaExtratoRepository.countByImportadaEmBetweenAndEstadoIn(j.de(), j.ate(),
                List.of(ConciliacaoService.SEM_CORRESPONDENCIA, ConciliacaoService.DIVERGENTE));
        long total = conciliadas + porResolver;
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("linhas", total);
        m.put("conciliadas", conciliadas);
        m.put("porResolver", porResolver);
        m.put("taxaAutomatica", total == 0 ? null : Math.round(conciliadas * 1000.0 / total) / 10.0);
        return m;
    }

    /** `dias` vem da query string: qualquer coisa que não seja um inteiro positivo cai nos 30 por omissão. */
    @Transactional(readOnly = true)
    public Map<String, Object> resumo(String dias) {
        int diasEfetivos = 30;
        if (dias != null && !dias.isBlank()) {
            try {
                int pedido = (int) Math.floor(Double.parseDouble(dias.trim()));
                if (pedido > 0) diasEfetivos = pedido;
            } catch (NumberFormatException ignored) {
                // "abc" → 30
            }
        }
        Janela periodo = janela(diasEfetivos, Instant.now());
        Map<String, Object> periodoOut = new LinkedHashMap<>();
        periodoOut.put("de", periodo.de());
        periodoOut.put("ate", periodo.ate());
        periodoOut.put("dias", diasEfetivos);
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("periodo", periodoOut);
        out.put("volume", volumeTransacionado(periodo));
        out.put("cotacoes", conversaoDeCotacoes(periodo));
        out.put("tempoAteConfirmacao", tempoAteConfirmacao(periodo));
        out.put("conciliacao", conciliacaoAutomatica(periodo));
        return out;
    }
}
